feat: Request, Job Hooks {Before/After}

gavindsouza opened the pull request:
This PR introduces [before, after] hooks for Requests & Background Jobs.

The hooks will be executed while the context of locals (request, response, flags, etc), the DB connection & such are maintained. 

## Request Hooks

These hooks allow you to run code before and after a request is made. These hooks can be defined as `list[str]` in which the function's dotted paths are defined.

The before hook is executed at the end of request initialization. This means accessing `frappe.request` in your `before_request` functions will be enough to get all relevant context.

The after hook is executed after the response object is built and database syncing &amp; cleanup tasks are completed, but before the DB connection is closed. In case of a request unsupported by the WSGI app, the response object may not exist. However, for most purposes accessing `frappe.response` should be enough. The request &amp; response objects are also passed into the `after_request` hooks if your function needs it.

Here are the request hooks that Frappe itself implements that you can find in it's codebase:

```python
before_request = [
	"frappe.recorder.record",
	"frappe.monitor.start",
	"frappe.rate_limiter.apply",
]
after_request = ["frappe.rate_limiter.update", "frappe.monitor.stop", "frappe.recorder.dump"]
```

## Background Job Hooks

These hooks allow you to run code before and after a background job is executed. These hooks can be defined as `list[str]` in which the function's dotted paths are defined.

The before and after hooks receive the job's dotted name `(method)`, kwargs `(kwargs)` and an extra kwarg (`transaction_type` for before_job, `result` for after_job), if your function will accept them.

Here are the job hooks that Frappe itself implements that you can find in it's codebase:

```python
before_job = [
	"frappe.monitor.start",
]
after_job = [
	"frappe.monitor.stop",
	"frappe.utils.file_lock.release_document_locks",
]
```

Docs: https://frappeframework.com/docs/v13/user/en/python-api/hooks/edit-wiki?wiki_page_patch=bcf13033cc

gavindsouza committed:
feat: Before/After Request Hooks

gavindsouza committed:
feat: Befor/After Job Hooks

gavindsouza committed:
test: Add tests for job sanity + hooks

gavindsouza committed:
test: Added test for before/after request hook

gavindsouza committed:
fix(background_jobs): Pass retval in execute_job

gavindsouza committed:
fix: Add request, job events in hooks boilerplate

gavindsouza committed:
fix: Pass result of job to after_job hooks

ankush commented on frappe/app.py:
```diff
@@ -318,7 +321,10 @@ def handle_exception(e):
 
 
-def after_request(rollback):
+def after_request(rollback: bool) -> bool:
```
should `after_requesst|job` hooks be run in reverse order? :thinking: 

Kinda makes sense for monitoring-type hooks.

Current "hooks" in app.py follow it. 
![image](https://user-images.githubusercontent.com/9079960/218428142-5711d667-e20e-4e05-9717-76c8042c577b.png)

netchampfaris replied:
@ankush 
> should after_requesst|job hooks be run in reverse order?

Makes sense but having the functions run in the defined order is easier to understand.

```
after_request = [
  'frappe.rate_limiter.update',
  'frappe.monitor.stop',
  'frappe.recorder.dump'
]
```

gavindsouza replied:
With @netchampfaris on this. The reverse order bit can be maintained by the respective application authors. If Frappe does the reversing, it won't be intuitive.

ankush commented on frappe/app.py:
```diff
@@ -318,7 +321,10 @@ def handle_exception(e):
-def after_request(rollback):
+def after_request(rollback: bool) -> bool:
+	for after_request_task in frappe.get_hooks("after_request"):
+		frappe.call(after_request_task)
```
`after_request` wont be called in case of exceptions. Maybe this should be in `finally` block?

netchampfaris commented:
@gavindsouza 

![image](https://user-images.githubusercontent.com/9355208/218762976-0bd29a99-27aa-4026-94cd-ad130d8f272d.png)

netchampfaris reviewed (commented):
The framework should also use the same API for monitor/record operations?

hooks.py
```py
before_request = [
  'frappe.rate_limiter.apply',
  'frappe.monitor.start',
  'frappe.recorder.record'
]
after_request = [
  'frappe.rate_limiter.update',
  'frappe.monitor.stop',
  'frappe.recorder.dump'
]
```

gavindsouza committed:
Merge branch 'develop' into request-job-hooks

gavindsouza committed:
fix(app): Move after_request hook inside finally block

Also, rename after_request fn to sync_database to better match functionality

gavindsouza committed:
refactor: Move before/after tasks as hooks

Moved before/after tasks in Requests as hooks for:
- monitor
- rate_limiter
- recorder

Moved before/after tasks in Jobs as hooks for:
- monitor
- releasing document locks

ankush commented:
Test failures seem related @gavindsouza reproducible locally too.

gavindsouza committed:
fix(after_hook): Don't pass exception object to hook

* it can fetch most relevant details via response object
* Exceptions not supported by Frappe's WSGI (unsupported HTTP methods) may not be accessible to the after_request hooks - but the lack of active response may be an indicator / and peeking in the request

gavindsouza committed:
Merge branch 'develop' into request-job-hooks

ascorbic-acid commented:
Thanks alot this is very useful this
